### PR TITLE
NOC initialization

### LINKED ISSUE #268: Connection to NOC should be initialized in `ttn`

### PR DESCRIPTION
**Pull request by rvolosatovs**
Closes #268

### CODE REVIEW

File: core/component.go [hunk 1]
```diff
@@ -118,6 +119,15 @@ func NewComponent(ctx log.Interface, serviceName string, announcedAddress string
 		go http.ListenAndServe(fmt.Sprintf(":%d", healthPort), nil)
 	}
 
+	if nocAddr := viper.GetString("noc-server"); len(nocAddr) > 0 {
+		conn, err := grpc.Dial(nocAddr, append(api.DialOptions, grpc.WithBlock(), grpc.WithInsecure())...)
+		if err != nil {
```
Comment (htdvisser) [inline core/component.go]:
```go
if err == nil {
    component.Monitor = pb_noc.NewMonitorClient(conn)
}
```
Things should not break when the NOC is unreachable

File: .env/handler/dev.yml [hunk 2]
```diff
@@ -8,3 +8,4 @@ tls: true
 key-dir: "./.env/handler/"
 
 auth-token: eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJ0dG4tYWNjb3VudC1zdGFnaW5nIiwic3ViIjoiZGV2IiwidHlwZSI6ImhhbmRsZXIiLCJpYXQiOjE0NzI4MTc5NTJ9.XyaTta7R66bfaCS4rzNYayzejE7nP7CcleksPyk71LsDkyMetFE_dJ-EP1Lp6Lys7yBILO4ax5tfuswBwvXg2sOIBIT30SPuTaJ62HidU6We_DQHBvHMSkdSOMnk3WNJ5rWhogga1rSHiKnJH56CqdcmT-9vwVFqIeU6ZCTawdDjRGjgQCn81PGlnCW7GVb3fLgpD1A_feFbGxBXVYSbRWA-rHGccyI9kWU5xxWyZwlYpl9_zT8j4mD8CZzffbgsPzag6aspE52fHgPRqz6rYzgZpXS7o6MUOEsjcWiOwp2ManvvnVA0mM_G65jdqMz1rPjdxt19QnQ8V59QtButGQ
+noc-server: noc.thethingsnetwork.org:1905
```
Comment (htdvisser) [inline .env/handler/dev.yml]: Don't put defaults in the config

File: .env/networkserver/dev.yml [hunk 3]
```diff
@@ -5,3 +5,4 @@ auth-servers:
   ttn-account-staging: "https://staging.account.thethingsnetwork.org"
 tls: true
 key-dir: "./.env/networkserver/"
+noc-server: noc.thethingsnetwork.org:1905
```
Comment (htdvisser) [inline .env/networkserver/dev.yml]: Don't put defaults in the config

File: .env/router/dev.yml [hunk 4]
```diff
@@ -8,3 +8,4 @@ tls: true
 key-dir: "./.env/router/"
 
 auth-token: eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJ0dG4tYWNjb3VudC1zdGFnaW5nIiwic3ViIjoiZGV2IiwidHlwZSI6InJvdXRlciIsImlhdCI6MTQ3MjgxNzk1MX0.Rjv2QHMAdhrZpc7Nf1HdUG1EoZqURr-PKWZaU7Xz7p30FMBxDXUDUtDbcILKbjstlczVOdHGQnAOf-rwzeIrXJ5051KopcDYcmpRwqJQHnKmzFClw5gs-1uAYSnwcEvpsGaFDS8Q7qc6X0m-5lFnFeHAtgVPa80joAw6QBPaTwW3KHvoY_a2htjk8tUQhVlL-wTCnnPMeUGIYJwHFIRXZW3Wk32XLWwF2w8fimVFgfr8yBQMLucbQ-CNEqcuC8tU8H46kPKCugadY6enmaOVVgeGFK1s9OEm4oHhYMJkE0agY7FZBPVEyl9bVxpCRor4H60MebA5U7xQ9tQQlP2ndQ
+noc-server: noc.thethingsnetwork.org:1905
```
Comment (htdvisser) [inline .env/router/dev.yml]: Don't put defaults in the config

File: core/component.go [hunk 5]
```diff
@@ -118,6 +119,15 @@ func NewComponent(ctx log.Interface, serviceName string, announcedAddress string
 		go http.ListenAndServe(fmt.Sprintf(":%d", healthPort), nil)
 	}
 
+	if nocAddr := viper.GetString("noc-server"); len(nocAddr) > 0 {
```
Comment (htdvisser) [inline core/component.go]: `nocAddr != ""`

File: cmd/root.go [hunk 6]
```diff
@@ -124,6 +131,9 @@ func init() {
 	RootCmd.PersistentFlags().String("discovery-server", "discover.thethingsnetwork.org:1900", "The address of the Discovery server")
 	viper.BindPFlag("discovery-server", RootCmd.PersistentFlags().Lookup("discovery-server"))
 
+	RootCmd.PersistentFlags().String("noc-server", "noc.thethingsnetwork.org:1905", "The address of the Noc server")
```
Comment (htdvisser) [inline cmd/root.go]: 1. of the **NOC** server
2. default should be `""` as NOC participation is opt-in

File: core/component.go (hunk 1 above)
Comment (rvolosatovs) [inline core/component.go]: 2b269bb
as discussed earlier in the office

File: .env/handler/dev.yml (hunk 2 above)
Comment (rvolosatovs) [inline .env/handler/dev.yml]: 0295694

File: .env/networkserver/dev.yml (hunk 3 above)
Comment (rvolosatovs) [inline .env/networkserver/dev.yml]: 0295694

File: .env/router/dev.yml (hunk 4 above)
Comment (rvolosatovs) [inline .env/router/dev.yml]: 0295694

File: core/component.go (hunk 5 above)
Comment (rvolosatovs) [inline core/component.go]: 9b6ac14

File: .env/broker/dev.yml [hunk 7]
```diff
@@ -12,3 +12,4 @@ auth-token: eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJ0dG4tYWNjb3VudC1zdGF
 broker:
   networkserver-cert: ./.env/broker/networkserver.cert
   networkserver-token: eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE0Njk3OTQzNTksIm5iZiI6MTQ2OTc5NDM1OSwic3ViIjoiZGV2In0.Xj1AFEAblzFLdS5cUiDrG773EqHazARvGdkTHFoPa_c0XGhdKSrevVpTNkzvcXeTQTWnTQrg1t98atNIk7F13Q
+noc-server: noc.thethingsnetwork.org:1905
```
Comment (rvolosatovs) [inline .env/broker/dev.yml]: 0295694

File: .env/discovery/dev.yml [hunk 8]
```diff
@@ -5,3 +5,4 @@ auth-servers:
   ttn-account: "https://account.thethingsnetwork.org"
   ttn-account-staging: "https://staging.account.thethingsnetwork.org"
 key-dir: "./.env/discovery/"
+noc-server: noc.thethingsnetwork.org:1905
```
Comment (rvolosatovs) [inline .env/discovery/dev.yml]: 0295694

File: core/component.go (hunk 1 above)
Comment (rvolosatovs) [inline core/component.go]: also 	5557a4f

File: core/component.go (hunk 5 above)
Comment (rvolosatovs) [inline core/component.go]: 6972800

File: cmd/root.go [hunk 9]
```diff
@@ -84,6 +84,12 @@ var RootCmd = &cobra.Command{
 			"Description":     viper.GetString("description"),
 			"DiscoveryServer": viper.GetString("discovery-server"),
 			"AuthServers":     viper.GetStringMapString("auth-servers"),
+			"NOCServer": func() string {
```
Comment (htdvisser) [inline cmd/root.go]: You can just do `"NOCServer": viper.GetString("noc-server"),`

Comment (rvolosatovs) [inline cmd/root.go]: Oh, yea, indeed.. needed that for the initial implementation, will change

File: cmd/root.go [hunk 10]
```diff
@@ -84,6 +84,7 @@ var RootCmd = &cobra.Command{
 			"Description":     viper.GetString("description"),
 			"DiscoveryServer": viper.GetString("discovery-server"),
 			"AuthServers":     viper.GetStringMapString("auth-servers"),
+			"NOCServer":       viper.GetString("noc-server"),
```
Comment (johanstokking) [inline cmd/root.go]: Can we have multiple Monitors as well?

This is the case when you have a simple monitor for a private network, but also want to report to the community NOC.

**Comment by rvolosatovs** [comment]
Closing for now, as there are issues with grpc connection

**Comment by rvolosatovs** [comment]
Shouldn't `api/noc` be renamed to `api/monitor` ?

**Pull request closed without merging**